src/python/pants/subsystem/subsystem.py

```diff
@@ -84,6 +84,8 @@ def is_initialized(cls) -> bool:
     return cls._options is not None
 
   # A cache of (cls, scope) -> the instance of cls tied to that scope.
+  # NB: it would be ideal to use `_S` rather than `Subsystem`, but we can't do this because
+  # MyPy complains that `_S` would not be properly constrained.
```

Comment:
Hrm, this still seems wanting or else calling out a typing bug of ours somewhere here.

Reply:
This is the full MyPy error:

> src/python/pants/subsystem/subsystem.py:89:38: error: Type variable
"pants.subsystem.subsystem._S" is unbound  [valid-type]
      _scoped_instances: Dict[Tuple[Type["_S"], str], "_S"] = {}
                                         ^
src/python/pants/subsystem/subsystem.py:89:38: note: (Hint: Use "Generic[_S]" or "Protocol[_S]" base class to bind "_S" inside a class)
src/python/pants/subsystem/subsystem.py:89:38: note: (Hint: Use "_S" in function signature to bind "_S" inside a function)

I don't think that either of their suggestions applies here. I'll improve the comment to make this more clear.

--

Also, this type hint isn't wrong, it's just less precise than it could be. MyPy respects Liskov substitution, i.e. this dict can hold any subclass of Subsystem. The reason we use the type var `_S` is so that the return type for functions is more specific, like `PythonSetup` rather than a generic `Subsystem`.